core/src/main/java/org/apache/accumulo/core/rpc/AccumuloProtocolFactory.java

```diff
@@ -0,0 +1,238 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.core.rpc;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.accumulo.core.trace.TraceUtil;
+import org.apache.thrift.TException;
+import org.apache.thrift.protocol.TCompactProtocol;
+import org.apache.thrift.protocol.TMessage;
+import org.apache.thrift.protocol.TProtocol;
+import org.apache.thrift.transport.TTransport;
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.context.propagation.TextMapGetter;
+
+/**
+ * Factory for creating instances of the AccumuloProtocol.
+ * <p>
+ * This protocol includes a custom header to ensure compatibility between different versions of the
+ * protocol.
+ */
+public class AccumuloProtocolFactory extends TCompactProtocol.Factory {
+
+  private static final long serialVersionUID = 1L;
+
+  private final boolean isClient;
+
+  public static class AccumuloProtocol extends TCompactProtocol {
+
+    static final int MAGIC_NUMBER = 0x41434355; // "ACCU" in ASCII
+    static final byte PROTOCOL_VERSION = 1;
+    private static final boolean HEADER_HAS_TRACE = true;
+
+    private final boolean isClient;
+
+    private Span span = null;
+    private Scope scope = null;
+    private final Map<String,String> traceHeaders = new HashMap<>();
+
+    public AccumuloProtocol(TTransport transport, boolean isClient) {
+      super(transport);
+      this.isClient = isClient;
+    }
+
+    /**
+     * For client calls, add RPC span and write the validation header
+     */
+    @Override
+    public void writeMessageBegin(TMessage message) throws TException {
+      if (!this.isClient) {
+        super.writeMessageBegin(message);
+      } else {
+        span = TraceUtil.startClientRpcSpan(this.getClass(), message.name);
+        scope = span.makeCurrent();
+
+        try {
+          this.writeHeader();
+          super.writeMessageBegin(message);
+        } catch (TException e) {
+          if (span != null) {
+            TraceUtil.setException(span, e, false);
+          }
+          if (scope != null) {
+            scope.close();
+          }
+          if (span != null) {
+            span.end();
+          }
+          throw e;
+        }
+      }
+    }
+
+    /**
+     * Writes the Accumulo protocol header containing version and identification info
+     */
+    private void writeHeader() throws TException {
+      super.writeI32(MAGIC_NUMBER);
+      super.writeByte(PROTOCOL_VERSION);
+
+      if (this.isClient && span != null && span.getSpanContext().isValid()) {
+        super.writeBool(HEADER_HAS_TRACE);
+        traceHeaders.clear();
+
+        W3CTraceContextPropagator.getInstance().inject(Context.current(), traceHeaders,
```

Comment:
Assuming this code was copied from TraceUtil?  Was there any significant change you needed to make to it to use it here?

Reply:
Yea it was pretty much just copied from TraceUtil.
```java
  public static TInfo traceInfo() {
    TInfo tinfo = new TInfo();
    W3CTraceContextPropagator.getInstance().inject(Context.current(), tinfo, TInfo::putToHeaders);
    return tinfo;
  }
```
The main difference is the use of the TInfo object instead of just a regular map in the new code.

Reply:
> I ended up refactoring out these inject and extract code blocks into methods in TraceUtil in https://github.com/apache/accumulo/commit/2c4eff6c50e9e1e07351400967ca6ab6ea2e88c8. Thats where they were before and I think its fine to keep them like that.

That is nice in that it makes the code in AccumuloProtocol more focused on its main task.